File: 788. rotated-digits.java
class Solution {
    public int rotatedDigits(int n) {
        int[] dp = new int[n + 1];
        int count = 0;

        for (int i = 0; i <= n; i++) {
            if (i < 10) {
                if (i == 0 || i == 1 || i == 8) {
                    dp[i] = 1;
                } else if (i == 2 || i == 5 || i == 6 || i == 9) {
                    dp[i] = 2;
                    count++;
                } else {
                    dp[i] = 0;
                }
            } else {
                int a = dp[i / 10];
                int b = dp[i % 10];

                if (a == 1 && b == 1) {
                    dp[i] = 1;
                } else if (a >= 1 && b >= 1) {
                    dp[i] = 2;
                    count++;
                } else {
                    dp[i] = 0;
                }
            }
        }

        return count;
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example 1
        int n1 = 10;
        int result1 = solution.rotatedDigits(n1);
        System.out.println("Example 1: Input: n = " + n1);
        System.out.println("Output: " + result1);
        System.out.println("Expected: 4\n");
        
        // Example 2
        int n2 = 1;
        int result2 = solution.rotatedDigits(n2);
        System.out.println("Example 2: Input: n = " + n2);
        System.out.println("Output: " + result2);
        System.out.println("Expected: 0\n");
        
        // Example 3
        int n3 = 2;
        int result3 = solution.rotatedDigits(n3);
        System.out.println("Example 3: Input: n = " + n3);
        System.out.println("Output: " + result3);
        System.out.println("Expected: 1");
    }
}

File: 788. rotated-digits.py
class Solution:
    def rotatedDigits(self, n: int) -> int:
        dp = [0] * (n + 1)
        count = 0

        for i in range(n + 1):
            if i < 10:
                if i in (0, 1, 8):
                    dp[i] = 1
                elif i in (2, 5, 6, 9):
                    dp[i] = 2
                    count += 1
                else:
                    dp[i] = 0
            else:
                a = dp[i // 10]
                b = dp[i % 10]

                if a == 1 and b == 1:
                    dp[i] = 1
                elif a >= 1 and b >= 1:
                    dp[i] = 2
                    count += 1
                else:
                    dp[i] = 0

        return count


# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example 1
    n1 = 10
    result1 = solution.rotatedDigits(n1)
    print(f"Example 1: Input: n = {n1}")
    print(f"Output: {result1}")
    print(f"Expected: 4\n")
    
    # Example 2
    n2 = 1
    result2 = solution.rotatedDigits(n2)
    print(f"Example 2: Input: n = {n2}")
    print(f"Output: {result2}")
    print(f"Expected: 0\n")
    
    # Example 3
    n3 = 2
    result3 = solution.rotatedDigits(n3)
    print(f"Example 3: Input: n = {n3}")
    print(f"Output: {result3}")
    print(f"Expected: 1")